docs/proposals/addon-auth-pattern.md

```diff
@@ -0,0 +1,150 @@
+- Do not use client certs in browsers, which are hard to install.
+
+The example used in this pattern is the kubernetes Dashboard.
+
+### Current pattern which has problems
+
+The following pattern is in use today.
+
+1. Dashboard is running with service `dashboard` in namespace `kube-system`.
+1. User runs `kubectl proxy -p 8001 &`.
+  -  This causes kubectl to serve HTTP on `localhost:8001`, and to reverse proxy that to the users current cluster, using TLS and the users
+     currently configured authentication method.
+1. User enters this into their browser navigation bar: `http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/#/workload`
+  - This causes the API server to proxy a single HTTP request.  The effect of the two proxies is that an HTTP request on localhost:8001 becomes an HTTP request for `/workloads` on port 80 on the pod IP of one of the Pods that backs services `dashboard` in namespace `kube-system`.  So far so good.
+1. The response is sent back from the Pod to the Apiserver.
+1. The response is sent back to the users.
+1. The user sees the Kubernetes Dashboard.
+
+The problem with this is that the Dashboard does not know which user is accessing it, so it cannot support multiple users with different capabilities.
+
+### Proposed Pattern
+
+The proposed pattern has the following key improvements over the current pattern:
+- a header is injected which tells the Dashboard pod the user who is reaching it.
+- the header can be validated easily as being signed by the apiserver, which is trusted to authenticate users
+
+1. Dashboard is running with service `dashboard` in namespace `kube-system`.
+1. User runs `kubectl proxy -p 8001 &`.
+  -  Same as before.
+1. User enters this into their browser navigation bar: `http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/?add-auth=true#/workload`
+  - Difference is the addition of the `add-auth=true` query parameter, which is handled by the APIserver.  When the API server sees this, it:
+    1. checks for a cached token
+    1. if not found, generates and signs a token which attests the identity of the user.
+    1. injects a header `Authorization: Bearer $TOKEN` into the HTTP request.
+1. The Dashboard sees the request.
+1. The dashboard sees the Authorization header and verifies the signature of the token, and extracts the username from one of the claims of the token.
+1. The dashboard uses the username as needed in its authorization logic (e.g. put into `Impersonate-User` header when doing actions on users behalf.
+1. The dashboard sends a response back to the user.
+1. The user sees the Kubernetes Dashboard.
+
+Properties of this token:
+- It is a JWT
+- The sub (subject) claim gives the kubernetes username of the user. The exact format of these is up to how the cluster admin setup apiserver auth in the cluster.
+- The exp (expiration) claim is some short amount of time, say a minute.  This prevents replay.
```

Comment:
I should not have said that `exp` prevents replay.  It just facilitates changes in access controls.    I updated the doc.  

If you trust your network and root users on your node, which I think a lot of people do, then I think this is useful.  I was trying to avoid solving the problem of distributing TLS certs to arbitrary servers.  We should totally solve that, and when we do, this solution will be safer to use with less-trusted networks.